Add schema inheritance support feature

DiHuynh35 opened the pull request:
# background

Add feature to support schema inheritance

DiHuynh35 committed:
Add schema inheritance support feature

hieuk09 commented on lib/open_api_2_json_schema.rb:
```diff
@@ -35,43 +37,51 @@ def self.build(type)
-        json_schema[struct] = convert_schema(data)
+        if struct == 'allOf' && data['$ref']
+          base_schema = JSON.parse(convert_from_file(data['$ref']))
+          base_schema = base_schema['schema'] if base_schema['schema']
```
In which case `schema` field does not exist in `base_schema`?

DiHuynh35 replied:
Hmmmmm actually i'm not sure, the sample code from swagger does not have `schema` in it, but I guess this gem is only for dealing with our specific case, so we can assume that `schema` presents all the time 👍

hieuk09 replied:
Let's not assume that the gem only deals with one specific case. 

So if `schema` field does not exists, we'll merge the whole hash, otherwise, we only merge the hash inside `schema`?

DiHuynh35 replied:
That's what the current logic is doing, right? Oh you meant something different? 🤔

hieuk09 replied:
No, I just want to confirm that is what we currently have.

Could you check what is the output of https://github.com/openapi-contrib/openapi-schema-to-json-schema and try to match it?

DiHuynh35 replied:
It seems like they doesn't support schema inheritance syntax either

<img width="1341" alt="image" src="https://user-images.githubusercontent.com/83625949/191475980-a260094d-d70d-43d9-8df0-cffc2e461e14.png">

<img width="446" alt="image" src="https://user-images.githubusercontent.com/83625949/191474940-89886676-cabe-4fbd-9d7f-c5b51ea2a738.png">

hieuk09 replied:
I think there is a note here: 

> NOTE: $refs are not handled in any way, so please use a resolver such as [json-schema-ref-parser](https://github.com/APIDevTools/json-schema-ref-parser) or [swagger-cli bundle](https://www.npmjs.com/package/swagger-cli) prior to using this package.

I guess to comply with the standard, we need to implement a ref parser similar to https://github.com/APIDevTools/json-schema-ref-parser, then in FE, we need to handle `allOf` logic instead. Could we check with Tim on whether handling `allOf` is possible on his side?

DiHuynh35 replied:
I see, just to confirm, the expected schema after parsing will be

```json
{
  "schema": {
    "allOf": {
      // <the merging between base schema and child schema>
    }
  }
}
``` 
Then, FE guys just need to skip the `allOf` and process its hash value as normal, right?

hieuk09 replied:
No, an example of the expected result is in https://github.com/hieuk09/open_api_2_json_schema/pull/1#discussion_r976462712. 

We should have this format:
```json
{
  "allOf": [
    {
      # schema 1 
    },
    {
      # schema 2
    }
  ]
}
```

For this gem, I think implementing the ref parser is enough. As I check the FE code, it's not easy to change to support `allOf`, so we can deep merge all schemas in `allOf` within the application instead.

hieuk09 commented on lib/open_api_2_json_schema.rb:
```diff
@@ -35,43 +37,51 @@ def self.build(type)
         end
       when Hash
-        json_schema[struct] = convert_schema(data)
+        if struct == 'allOf' && data['$ref']
```
Can we handle this `allOf` outside of this loop?

hieuk09 reviewed (commented):
Could we separate the styling fix with the actual fix so that it's easier to review?

DiHuynh35 committed:
Revert "Add schema inheritance support feature"

This reverts commit ec1a507e6fba617b1e511b7a0662fa446cb570a6.

hieuk09 commented on spec/fixtures/inherited_json_schema.json:
```diff
@@ -0,0 +1,78 @@
+{
+  "schema": {
```
To comply with JSON schema standard, I think the result of the parser should be:

```json
{
  "allOf":[
    {
      "type":"object",
      "required":[
        "personal_details",
        "login_details",
        "agreements"
      ],
      "properties":{
        "personal_details":{
          "type":"object",
          "properties":{
            "first_name":{
              "type":"string"
            },
            "last_name":{
              "type":"string"
            },
            "birthdate":{
              "type":"string",
              "format":"date"
            },
            "email":{
              "type":"string",
              "format":"email"
            },
            "email_confirmation":{
              "type":"string",
              "format":"email"
            }
          },
          "required":[
            "first_name",
            "last_name",
            "birthdate",
            "email",
            "email_confirmation"
          ]
        }
      },
      "login_details":{
        "type":"object",
        "properties":{
          "star_alliance_id":{
            "type":"string",
            "pattern":"^[0-9]{10}$",
            "format":"id"
          }
        },
        "required":[
          "star_alliance_id"
        ]
      },
      "agreements":{
        "type":"object",
        "required":[
          "terms"
        ],
        "properties":{
          "terms":null,
          "type":"boolean",
          "x-display-type":"checkbox"
        }
      }
    },
    {
      "type":"object",
      "properties":{
        "login_details":{
          "type":"object",
          "required":[
            "star_alliance_id",
            "password",
            "password_confirmation"
          ],
          "properties":{
            "star_alliance_id":{
              "type":"string",
              "pattern":"^[0-9]{10}$",
              "format":"id"
            },
            "password":{
              "type":"string",
              "format":"password"
            },
            "password_confirmation":{
              "type":"string",
              "format":"password"
            }
          }
        }
      }
    }
  ]
}
```

hieuk09 replied:
You can check whether the form can be rendered correctly in https://rjsf-team.github.io/react-jsonschema-form/

DiHuynh35 committed:
Add schema inheritance support feature

DiHuynh35 commented:
@hieuk09 I've just addressed your comments, could you take a look? 🙇 

The schema seems to be rendered correctly in `react-jsonschema-form`
![image](https://user-images.githubusercontent.com/83625949/191660025-eb9068e3-3943-4b5e-bf5f-a9f58a38a2d3.png)

I will check with AP team whether the library they're using can support this, if not, we can merge all schemas from `allOf` in BE before returning to FE as you mentioned

### caveat

For the `ref schema parser`, currently I'm doing a hack to skip the key `schema` from the loaded `$ref` schema using `move_hash_value_one_upper_level`, cuz the form can't be rendered with it. Not sure if this is the right way to handle such the case

![image](https://user-images.githubusercontent.com/83625949/191660898-711b1ea8-6a27-4313-b034-6401f5ee4941.png)

hieuk09 commented on spec/fixtures/child.yml:
```diff
@@ -0,0 +1,24 @@
+schema:
+    allOf:
+      $ref: 'spec/fixtures/father.yml'
+      type: object
+      required:
+        - login_details
+      properties:
+        login_details:
+          type: object
+          required:
+            - star_alliance_id
+            - password
+            - password_confirmation
+          properties:
+            star_alliance_id:
+              type: string
+              pattern: ^[0-9]{10}$
+              format: id
+            password:
+              type: string
+              format: password
+            password_confirmation:
+              type: string
+              format: password
```
This is not valid `allOf` syntax, it should be:
```suggestion
      - $ref: 'spec/fixtures/father.yml'
      - type: object
        required:
          - login_details
        properties:
          login_details:
            type: object
            required:
              - star_alliance_id
              - password
              - password_confirmation
            properties:
              star_alliance_id:
                type: string
                pattern: ^[0-9]{10}$
                format: id
              password:
                type: string
                format: password
              password_confirmation:
                type: string
                format: password
```

DiHuynh35 replied:
🤦‍♂️

hieuk09 commented on lib/open_api_2_json_schema.rb:
```diff
@@ -60,6 +64,14 @@ def convert_schema(schema)
       end
     end
 
+    if schema.key?('allOf') && schema['allOf'].key?('$ref')
```
Now that we don't merge schema anymore, no need to implement this field separately from the loop.

hieuk09 commented on lib/open_api_2_json_schema.rb:
```diff
@@ -147,4 +164,12 @@ def clean_required(required, properties)
     properties.slice(*required).keys
   end
+
+  def move_hash_value_one_upper_level(hash, upper_key)
```
Do we need this?

DiHuynh35 replied:
The ref-schema might have the top-level `schema` key, which won't work for [this syntax](https://github.com/hieuk09/open_api_2_json_schema/pull/1#discussion_r976584180) when trying on `react-jsonschema-form`

DiHuynh35 replied:
Same as https://github.com/Kaligo/guardhouse/pull/4837#issuecomment-1254766084

hieuk09 reviewed (approved):
LGTM.
